src/python/pants/engine/fs_test.py

```diff
@@ -1502,3 +1507,45 @@ def test_snapshot_diff(
     assert diff.their_unique_files == expected_diff.our_unique_files
     assert diff.their_unique_dirs == expected_diff.our_unique_dirs
     assert diff.changed_files == expected_diff.changed_files
+
+
+def test_path_metadata_request(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "foo": b"xyzzy",
+            "sub-dir/bar": b"12345",
+        }
+    )
+    os.symlink("foo", os.path.join(rule_runner.build_root, "bar"))
+
+    def get_metadata(path: str) -> PathMetadata | None:
+        result = rule_runner.request(PathMetadataResult, [PathMetadataRequest(path)])
+        print(f"metadata={result.metadata}")
+        return result.metadata
+
+    m1 = get_metadata("foo")
+    assert m1 is not None
+    assert m1.path == "foo"
+    assert m1.kind == PathMetadataKind.FILE.value
+    assert m1.length == len(b"xyzzy")
+    assert m1.symlink_target is None
+
+    m2 = get_metadata("not-found")
+    assert m2 is None
+    (Path(rule_runner.build_root) / "not-found").write_bytes(b"is found")
+    time.sleep(0.250)
```

Comment:
Sleep is necessary to allow filesystem watcher to invalidate the graph node.

Reply:
Where did 250ms come from? Is there a more principled way to wait for the watcher?

Reply:
Pulled out of thin air. It was sufficient delay for the filesystem watcher to trigger, but is probably not the minimal delay which would suffice.

A more principled way would be for there to be something like a "barrier" which the test could await until all filesystem change events up to that point have propagated. I do not believe we have anything like that.

The code could also just retry N times until the change event has propagated, but that is only slightly more principled than just a plain old delay.

Reply:
I added a helper function to do the suggested retries.